internal/internal_worker.go

```diff
@@ -1210,6 +1211,23 @@ func (aw *WorkflowReplayer) ReplayWorkflowExecution(ctx context.Context, service
 	return aw.replayWorkflowHistory(logger, service, namespace, &history)
 }
 
+// inferInvocations extracts the set of *interactionpb.Invocation objects that
+// should be attached to a workflow task (i.e. the
+// PollWorkflowTaskQueueResponse.Interactions) if that task were to carry the
+// provided slice of history events.
+func inferInvocations(events []*historypb.HistoryEvent) []*interactionpb.Invocation {
+	var invocations []*interactionpb.Invocation
+	for _, e := range events {
+		if attrs := e.GetWorkflowUpdateAcceptedEventAttributes(); attrs != nil {
```

Comment:
An update is always accepted or rejected as part of the workflow task processing on the task that carries the initial request. Rejections are (for now) never made durable so the only evidence that an update was ever invoked is the AcceptedEvent. Since this function is inferring the existence of invocations that's what we check here.

Reply:
Okay, I went over the protocol and realized that error returned from the update handler ends up as completed with a failure result variant.

So I guess we could save the extra event when the handler completes immediately but that's just an optimization and is probably not worth it right now.

Reply:
Yes, in this discussion in the past we decided it was preferable to always have the accepted event, even if it could be implied in some instances by the completed event.